### DISCUSSION TITLE
[Feature]: Support native deployments (without docker, includes example)

### DISCUSSION THREAD

**Discussion by oddlama** [immich-app/immich#1657] [Feature Request]
### Feature detail

Thanks for the great app! One thing that I'm dearly missing is the ability to deploy this on bare-metal, without relying on docker. The problem is that in certain situations, docker might not be a viable option to deploy a project. I for example am using an extremely strict nftables firewall, for which docker simply has no compatibility (and docker will likely still take some years to catch up). Furthermore, I deploy each application in its own KVM virtual machine with libvirt, so docker would just be an unnecessary overhead for me.

I have already managed to get immich running without docker just fine, it would just be missing a tiny bit of packaging to make the process more accessible and repeatable. I've attached my scripts to build and deploy immich below in case you (or any other people stumbling over this) are interested.

## Installing on bare-metal

1. Create user for immich (`/var/lib/immich`)
2. Create environment file `/var/lib/immich/env` (adapt from upstream provided `.env.example`)
3. Build app and create links to upload folder with script `runuser -u immich -- ./build.sh`
4. Map container aliases to localhost in /etc/hosts like this: `127.0.0.1       localhost immich-server immich-microservices immich-machine-learning immich-web`
5. Create fstab entry to bind mount `/the/backup/folder/you/want` to `/var/lib/immich/upload`
6. Start redis, postgresql and provided systemd services

The need to map the upload folder in several app directories is a little clunky right now, just like having to use /etc/hosts because hostnames are hardcoded. But the rest is working like a charm without docker.

You might need to change the systemd services a little to fit your system if e.g. your postgresql service is not called `postgresql-14.service`, or when it is running on another system. I've also used systemd to do sandboxing for the services, so they will not be able to access other directories if you need that for some reason.

### Buildscript `build.sh`

This script is adapted from the Dockerfiles I found in this repo.

```bash
#!/bin/bash

IMMICH_TAG=v1.29.4_44-dev

set -uo pipefail
function die() { echo "error: $*" >&2; exit 1; }
function try() { "$@" || die "Command failed: $*"; }

TMP="$(mktemp --tmpdir=/var/tmp -d)" || die "Could not create tmpdir"
trap 'rm -rf -- "$TMP"' EXIT

try cd "$TMP"
try git clone https://github.com/immich-app/immich
try cd "$TMP"/immich
try git checkout "$IMMICH_TAG"

# Build server
try cd "$TMP"/immich/server
try npm ci
try npm run build
try npm prune --production
try mkdir app
try cp -a package.json package-lock.json node_modules dist app

# Build web frontend
try cd "$TMP"/immich/web
try npm ci
try npm run build
# TODO: why is the whole folder used here? Can we strip something from here? What about npm prune --production?

# Build machine-learning
try cd "$TMP"/immich/machine-learning
try npm ci
try npm rebuild @tensorflow/tfjs-node --build-from-source
try npm run build
try npm prune --production
try mkdir app
try cp -a package.json package-lock.json node_modules dist app

# Copy only relevant files to new directory
try cd "$TMP"/
try mkdir app
try mv "$TMP"/immich/server/app app/server
try mv "$TMP"/immich/web app/web
try mv "$TMP"/immich/machine-learning/app app/machine-learning

# Install new files (deleting old deployment)
[[ -e /var/lib/immich/app ]] && try rm -rf /var/lib/immich/app
try cp -r app /var/lib/immich/app

# Ensure the upload folder is available
mkdir -p /var/lib/immich/upload
ln -s /var/lib/immich/upload /var/lib/immich/app/server/upload
ln -s /var/lib/immich/upload /var/lib/immich/app/machine-learning/upload
```

### Systemd service `immich-server.service`

```systemd.service
[Unit]
Description=immich server
Documentation=https://github.com/immich-app/immich
Requires=redis-immich.service
Requires=postgresql-14.service
After=network.target
Wants=network-online.target
After=network-online.target

[Service]
User=immich
Group=immich
Type=simple
Restart=on-failure

EnvironmentFile=/var/lib/immich/env
Environment=NODE_ENV=production
SyslogIdentifier=immich-server
ExecStart=node dist/apps/immich/apps/immich/src/main

PrivateDevices=true
ProtectHome=true
ProtectSystem=strict
ProtectKernelTunables=yes
ProtectKernelModules=yes
ProtectControlGroups=yes

RestrictNamespaces=yes

SystemCallArchitectures=native
AmbientCapabilities=
CapabilityBoundingSet=
NoNewPrivileges=yes

WorkingDirectory=/var/lib/immich/app/server
ReadWritePaths=/tmp /var/tmp /var/lib/immich/upload

[Install]
WantedBy=multi-user.target
```

### Systemd service `immich-microservices.service`

```systemd.service
[Unit]
Description=immich microservices
Documentation=https://github.com/immich-app/immich
Requires=redis-immich.service
Requires=postgresql-14.service
After=network.target
Wants=network-online.target
After=network-online.target

[Service]
User=immich
Group=immich
Type=simple
Restart=on-failure

EnvironmentFile=/var/lib/immich/env
Environment=NODE_ENV=production
SyslogIdentifier=immich-microservices
ExecStart=node dist/apps/microservices/apps/microservices/src/main

PrivateDevices=true
ProtectHome=true
ProtectSystem=strict
ProtectKernelTunables=yes
ProtectKernelModules=yes
ProtectControlGroups=yes

RestrictNamespaces=yes

SystemCallArchitectures=native
AmbientCapabilities=
CapabilityBoundingSet=
NoNewPrivileges=yes

WorkingDirectory=/var/lib/immich/app/server
ReadWritePaths=/tmp /var/tmp /var/lib/immich/upload

[Install]
WantedBy=multi-user.target
```

### Systemd service `immich-machine-learning.service`

```systemd.service
[Unit]
Description=immich machine learning
Documentation=https://github.com/immich-app/immich
Requires=postgresql-14.service
After=network.target
Wants=network-online.target
After=network-online.target

[Service]
User=immich
Group=immich
Type=simple
Restart=on-failure

EnvironmentFile=/var/lib/immich/env
Environment=NODE_ENV=production
SyslogIdentifier=immich-machine-learning
ExecStart=node dist/main.js

PrivateDevices=true
ProtectHome=true
ProtectSystem=strict
ProtectKernelTunables=yes
ProtectKernelModules=yes
ProtectControlGroups=yes

RestrictNamespaces=yes

SystemCallArchitectures=native
AmbientCapabilities=
CapabilityBoundingSet=
NoNewPrivileges=yes

WorkingDirectory=/var/lib/immich/app/machine-learning
ReadWritePaths=/tmp /var/tmp /var/lib/immich/upload

[Install]
WantedBy=multi-user.target
```

### Systemd service `immich-web.service`

```systemd.service
[Unit]
Description=immich web
Documentation=https://github.com/immich-app/immich
After=network.target
Wants=network-online.target
After=network-online.target

[Service]
User=immich
Group=immich
Type=simple
Restart=on-failure

EnvironmentFile=/var/lib/immich/env
Environment=NODE_ENV=production
SyslogIdentifier=immich-web
ExecStart=node build/index.js

PrivateDevices=true
ProtectHome=true
ProtectSystem=strict
ProtectKernelTunables=yes
ProtectKernelModules=yes
ProtectControlGroups=yes

RestrictNamespaces=yes

SystemCallArchitectures=native
AmbientCapabilities=
CapabilityBoundingSet=
NoNewPrivileges=yes

WorkingDirectory=/var/lib/immich/app/web
ReadWritePaths=/tmp /var/tmp

[Install]
WantedBy=multi-user.target
```

---

P.S.: Is it intended that compiling the most recent release with npm reports 10+ high and 3 critical security issues? Not sure what to do about that as a user.

### Platform

Server

**Comment by StillLoading** [comment]
Thank you for this. Really appreciate the work you put into sorting this out. 

For anyone wanting to deploy on OpenSUSE Tumbleweed/MicroOS here are some pointers.

## Dependencies
For building
```
make gcc gcc-c++ npm18 nodejs18 nodejs18-devel nodejs-common
```
For hosting
```
nginx redis postgresql-server
```
## SELinux
Allow local nginx proxy to connect to local services via port
```
setsebool -P httpd_can_network_connect 1
```
## Nginx
Grab the config from this repository and modify to your needs.

**Comment by huangfengjing** [comment]
Nice work.
Can anybody compose a Batch-Script/PowerShell script like this to help deploying immich on Windows?

**Comment by jumbi77** [comment]
Thanks a lot.
One question: Is this `Requires=redis-immich.service` correct in systemd.services? My system cant find those and i had to change them `Requires=redis-server.service`

**Comment by StillLoading** [comment]
@jumbi77 I think @oddlama did a good job at pointing that out in his post.

> You might need to change the systemd services a little to fit your system if e.g. your postgresql service is not called `postgresql-14.service`, or when it is running on another system...

**Comment by hydazz** [comment]
(off topic)
using this installing on bare-metal method, I am creating a (hopefully) functioning single immich container for me to use on unraid (docker-compose is really messy). main issue is that the server expects hostnames `immich-server` `immich-microservices` `immich-machine-learning` `immich-web` and the `hosts` file cant be modified on docker so I end up with

`Error: getaddrinfo ENOTFOUND immich-server`

https://github.com/hydazz/docker-immich
#appsshouldbeasignecontainerwhereapplicable

edit: all seems to be working in latest commit...

**Comment by alextran1502** [comment]
> (off topic) using this installing on bare-metal method, I am creating a (hopefully) functioning single immich container for me to use on unraid (docker-compose is really messy). main issue is that the server expects hostnames `immich-server` `immich-microservices` `immich-machine-learning` `immich-web` and the `hosts` file cant be modified on docker so I end up with
> 
> `Error: getaddrinfo ENOTFOUND immich-server`
> 
> https://github.com/hydazz/docker-immich #appsshouldbeasignecontainerwhereapplicable
> 
> edit: all seems to be working in latest commit...

Thank you for the contribution. 

I have a thought about your hashtag #appsshouldbeasignecontainerwhereapplicable

Let's say you want to limit the CPU of `machine-learning` and `microservices` containers which handle the intensive background tasks, so that your core server can still handle I/O operations effectively and efficiently while there are people performing backing up. Can you do that with the single container image?

**Comment by hydazz** [comment]
> Let's say you want to limit the CPU of `machine-learning` and `microservices` containers which handle the intensive background tasks, so that your core server can still handle I/O operations effectively and efficiently while there are people performing backing up. Can you do that with the single container image?

Yes, that's not really possible on a single container - there are indeed use cases for having the containers separated on lower-powered hardware. Most self-hosted image services have single containers which makes deployment much easier on specific OSes (ie Unraid), where most users won't even think about limiting the CPU of a container.

**Comment by bt90** [comment]
You could try to start the container with the `sys_nice` capability and change the process priorities inside the container to accomplish that.

https://docs.docker.com/config/containers/resource_constraints/#configure-individual-containers

Pair that with e.g dumb-init:

https://github.com/Yelp/dumb-init#session-behavior

**Comment by hydazz** [comment]
> You could try to start the container with the `sys_nice` capability and change the process priorities inside the container to accomplish that.
> 
> https://docs.docker.com/config/containers/resource_constraints/#configure-individual-containers
> 
> Pair that with e.g dumb-init:
> 
> https://github.com/Yelp/dumb-init#session-behavior

there should be a way to achieve this with s6 and s6-overlay

Why does the machine-learning module spawn 40+ processes that use up over 1GB of ram at idle? is this necessary? is it possible to have the machine-learning run on-demand rather than all the time?
![image](https://user-images.githubusercontent.com/53986978/210122124-ac144366-bf82-48d9-96b5-d46a5896c2f5.png)

**Comment by denis-ali** [comment]
Any chance on updating this to v1.50.x?

Seems since 1.48 some changes were introduced, like the machine learning module in python, which no longer makes the script that much valid.

**Comment by derzahla** [comment]
awesome. thanks for this.  Im trying to put together immich in a freebsd jail on truenas so this will be helpful

**Reply by jeroenh** [reply]
Happy to help with this, I'd like to run this on my FreeBSD server as well.

**Reply by spusuf** [reply]
Same with me, did you get it working on BSD?

**Reply by arcepi** [reply]
I have it working in a FreeBSD 13.1 jail. I followed more or less the Dockerfiles to adapt the process and also created rc.d scripts.

The only thing I didn't get working is the machine-learning part as it requires pytorch2, which doesn't have a port for FreeBSD yet.

I want something more automated so I began today to try to make ports for immich_web, immich_server and immich_microservices, but it is my first time writing a FreeBSD port so progress is slow. 
I already have a port for immich_web which I think installs the web part correctly in /usr/local/www/immich/web and its rc.d file.

The main problem I had with my "manual" deployment is that sometimes the applications don't use the values set in the environment file (the most important that doesn't work is the path to the stored files). But it is probably my fault.

If I manage to have the ports working I'll put a comment here and I will try sending them to FreeBSD ports (although I don't know how involved that process will be).

**Reply by spusuf** [reply]
I'd definitely be keen to replicate that. Maybe make a GitHub page for all the files, install script (any commands you used in the console), and changes needed (such as environment variables). I can contribute on there.

**Reply by arcepi** [reply]
Hello! I can't send you the steps I did because I'm on holidays and I have them on my desktop computer.

Anyway, I have ready the ports for the web, server and microservices. With that I've been running Immich (although I know it is limited).
[0001-Add-changes-for-new-ports-for-Immich.patch](https://github.com/immich-app/immich/files/12371679/0001-Add-changes-for-new-ports-for-Immich.patch)

I attach a git diff to the FreeBSD ports tree. Please, test the two ports. If I did everything right it should be as simple as doing make && make install and then editing environment files in /usr/local/etc/immich to fit your environment and configuring nginx. For that I also install a template in /usr/local/etc/nginx.

Hope it works.

**Comment by tgutzler** [comment]
This is going to be a lengthy discussion, I think...
Following the OP's efforts, I'm attempting to run immich on a proxmox container running debian bookworm. However, things have changed and the build script no longer runs, so I'm working my way through the Dockerfiles myself.
I checked out master on my build server, built the server part, copied the previously mentioned files across to the container, installed a bunch of dependencies and am now trying to get the microservices up. However, I'm getting an error from the redis client ioredis:

> node dist/main microservices
[Nest] 10643  - 08/20/2023, 10:42:22 AM     LOG [NestFactory] Starting Nest application...
[Nest] 10643  - 08/20/2023, 10:42:22 AM     LOG [InstanceLoader] TypeOrmModule dependencies initialized +60ms
[Nest] 10643  - 08/20/2023, 10:42:22 AM     LOG [InstanceLoader] BullModule dependencies initialized +0ms
[Nest] 10643  - 08/20/2023, 10:42:22 AM     LOG [InstanceLoader] ConfigHostModule dependencies initialized +2ms
[Nest] 10643  - 08/20/2023, 10:42:22 AM     LOG [InstanceLoader] DiscoveryModule dependencies initialized +0ms
[Nest] 10643  - 08/20/2023, 10:42:22 AM     LOG [InstanceLoader] ConfigModule dependencies initialized +9ms
[Nest] 10643  - 08/20/2023, 10:42:22 AM     LOG [InstanceLoader] BullModule dependencies initialized +0ms
[Nest] 10643  - 08/20/2023, 10:42:22 AM     LOG [InstanceLoader] BullModule dependencies initialized +0ms
[Nest] 10643  - 08/20/2023, 10:42:23 AM     LOG [InstanceLoader] TypeOrmCoreModule dependencies initialized +211ms
[Nest] 10643  - 08/20/2023, 10:42:23 AM     LOG [InstanceLoader] TypeOrmModule dependencies initialized +1ms
[Nest] 10643  - 08/20/2023, 10:42:23 AM     LOG [InstanceLoader] TypeOrmModule dependencies initialized +0ms
[Nest] 10643  - 08/20/2023, 10:42:23 AM     LOG [InstanceLoader] InfraModule dependencies initialized +9ms
[Nest] 10643  - 08/20/2023, 10:42:23 AM     LOG [InstanceLoader] DomainModule dependencies initialized +5ms
[Nest] 10643  - 08/20/2023, 10:42:23 AM     LOG [InstanceLoader] MicroservicesModule dependencies initialized +0ms
[Nest] 10643  - 08/20/2023, 10:42:23 AM     LOG [MetadataExtractionProcessor] Reverse geocoding is enabled
[Nest] 10643  - 08/20/2023, 10:42:23 AM     LOG [MetadataExtractionProcessor] Initializing Reverse Geocoding
Error: connect ETIMEDOUT
    at Socket.<anonymous> (/var/lib/immich/app/server/node_modules/ioredis/built/Redis.js:170:41)
    at Object.onceWrapper (node:events:627:28)
    at Socket.emit (node:events:513:28)
    at Socket._onTimeout (node:net:569:8)
    at listOnTimeout (node:internal/timers:564:17)
    at process.processTimers (node:internal/timers:507:7) {
  errorno: 'ETIMEDOUT',
  code: 'ETIMEDOUT',
  syscall: 'connect'
}

My redis appears to be running:
> systemctl status redis
● redis-server.service - Advanced key-value store
     Loaded: loaded (/lib/systemd/system/redis-server.service; enabled; preset: enabled)
     Active: active (running) since Sat 2023-08-19 21:05:31 AWST; 13h ago
       Docs: http://redis.io/documentation,
             man:redis-server(1)
   Main PID: 5464 (redis-server)
     Status: "Ready to accept connections"
      Tasks: 5 (limit: 9304)
     Memory: 5.2M
        CPU: 1min 2.470s
     CGroup: /system.slice/redis-server.service
             └─5464 "/usr/bin/redis-server 127.0.0.1:6379"
                                                                                                                        Aug 19 21:05:31 immich systemd[1]: Starting redis-server.service - Advanced key-value store...
Aug 19 21:05:31 immich systemd[1]: Started redis-server.service - Advanced key-value store.

I have added the following entry to my /etc/hosts file:
127.0.0.1 immich_redis immich_postgres typesense immich-server immich-microservices immich-machine-learning immich-web

What am I missing?

**Comment by timdonovanuk** [comment]
Please add native (bare metal) install support. :/

**Reply by alextran1502** [reply]
You can follow the steps in the docker file for each service (server, web, proxy) to replicate the native installation.

**Reply by Aerodin** [reply]
You may want to take a look at this thread:
https://github.com/tteck/Proxmox/discussions/1096

A Debian install path on bare metal would allow someone like TTeck to create a simple install script to install it on Proxmox or an LXC container which would lead to more widespread adoption.  Perhaps a collaboration between the Immich team and TTek can lead to him creating a simple install script for Proxmox/LXC which could be leveraged by other platforms like Unraid etc. that are seeking to host this in a non-Docker native environment as well.  Then everyone benefits from a simplified installation process.

I am watching this software with fascination and want to try it out, however, the comments above about issues installing it bare metal copying the docker files scares me away.

**Reply by timdonovanuk** [reply]
Yep! Huge +1 to this comment. I'm a huge fan of docker, and started my self hosting journey using it heavily. I've since moved over to LXC's for everything, and indeed a 80% of the ~ 30 LXC's I run come from tteck. I was disappointed to see there is no immich lxc on the tteck site, so was looking into why, and that is how I ended up here asking for bare metal install support. Completely agree that a tteck script would really help drive adoption of immich.

**Reply by anubis-genix** [reply]
This would be an enormous enrichment for the non-technical users!

**Reply by tgutzler** [reply]
Sign me up for beta-testing LXCs!

**Reply by SR-Lut3t1um** [reply]
I guess it would be best if someone would contribute build scripts, to build the packages in apt/rpm. Than a simple repo could be provided making installing on bare metal a lot simpler. Postgres, redis etc. can be expected to be installed and configured manually, however a little work for documentation would be necessary. I don't know what the current state of the art of package building looks like, however something like [OBS](https://openbuildservice.org/help/manuals/obs-user-guide/art.obs.bg) might help here, (vaultwarden uses this approach too).

**Comment by dpantel** [comment]
Slightly off-topic:

1. I stumbled on [this project](https://github.com/nsg/immich-distribution) the other day. It’s an unofficial Snap repackaging. It’s not going to help any of us wanting to run “bare-metal”, but might be a good option for Ubuntu users instead of using Docker. 

2. I’m successfully running Docker/Immich in an unprivileged LXC container on Proxmox. The only concession I had to make was a raw ext4 virtual disk for the Docker directory. That’s only necessary because my backing FS is ZFS. And on that front, according to recent commits, ZFS might now/soon support unprivileged container overlay FS, which might allow running Docker in LXC on ZFS with no issues.

**Reply by steadfasterX** [reply]
Nice i have the same setup (proxmox + zfs). I haven't looked into immich yet so is it easy to set it up that way?

**Comment by smiba** [comment]
Not having native deployments is incredibly unfortunate, I wish I could just install this inside an LXC container without docker

**Reply by micash545** [reply]
I found a method but I'm still working on it. I will share soon

**Reply by steadfasterX** [reply]
> I found a method but I'm still working on it. I will share soon

Any news on that topic?

**Reply by Ondjultomte** [reply]
> Not having native deployments is incredibly unfortunate, I wish I could just install this inside an LXC container without docker

I concur! "native" install support is highly needed

**Comment by arter97** [comment]
Seems like Immich changed a quite since the last time OP was updated.

I took the liberty of creating some scripts to natively install Immich v1.97.0.

https://github.com/arter97/immich-native

Leaving it here so that other people can find it useful.

**Reply by arcepi** [reply]
Very useful! I have almost ready port files for FreeBSD and the only piece missing was pgvecto.rs, but I see that using pgvector is also compatible!!!

Do you know if there is any limitation when using pgvector instead of pgvecto.rs?

**Reply by mmomjian** [reply]
You can read more about pgvector [here](https://github.com/immich-app/immich/releases/tag/v1.95.0). Currently, the answer is "some features may not work quite as well with pgvector". The devs primarily operate with vecto.rs so I doubt there is much testing done with pgvector.

**Reply by arcepi** [reply]
Thanks @mmomjian, I had completely missed that part about pgvector compatibility.

**Reply by daftaupe** [reply]
> Very useful! I have almost ready port files for FreeBSD and the only piece missing was pgvecto.rs, but I see that using pgvector is also compatible!!!

@arcepi Very interesting, I have immich running on FreeBSD. I can give your port a try/test whenever you feel it's ready.

**Reply by Ondjultomte** [reply]
> Seems like Immich changed a quite since the last time OP was updated.
> 
> I took the liberty of creating some scripts to natively install Immich v1.97.0.
> 
> https://github.com/arter97/immich-native
> 
> Leaving it here so that other people can find it useful.

Hey! thanx for your work unfortenetly that script fails when I run it in debian 12 lxc

**Comment by PedanticAvenger** [comment]
I'm adding a vote to the "Please provide native deployment support".  From my perspective it can be leveraged into a docker container build workflow and allow LXC container support (or other bare-metal installs) without increasing support/deployment load meaningfully.

**Comment by exstntlmsnthrp** [comment]
I've heard so many good things about immich and have finally decided to try it out. Then come and find a discussion like this and see that docker is still the only official install method and people are still faffing about to get it to work on bare metal. 

That's just so disappointing. 😞

**Reply by alextran1502** [reply]
In order to support different RAW formats, we have to compile some libraries a specific way, they are documented in the Docker file, so if we have to support native deployment, we have to find ways to make it work across different Linux environments. That is the trade off of time we can use to focus on developing the application instead.

**Reply by exstntlmsnthrp** [reply]
Thanks for taking the time to respond and for your explanation. Much appreciated. I understand every developer has to make project decisions and this is what was decided was best here.

**Comment by 4v3ngR** [comment]
I'd like to give a huge shout of thanks to [arter97](https://github.com/arter97) for his install script, for I'm a "macOS can be a server" user (I've an old mac mini as a server) and with his script (and a few changes) I was able to get immich installed and running on macOS monterey and Sonoma.

I've also updated my script (located [here](https://github.com/4v3ngR/immich-native-macos)) to do full end to end installation. It can now:
- install external dependencies (via brew)
- create the immich user and group
- initialise the postresql database
- build and install immich
- setup the launchd scripts

So, we just need a "Windows can be a server" user to create a Windows install script and we'll have unofficial bare metal installers on all systems.

**Reply by blublub** [reply]
Hello everyone,

This is just my 2cents:

If Immich wants to gain market share beyond "power users" it needs an easy way to install on Linux, Windows and Mac OS - the latter two being the most important with a combined market share around 90% - or in other words: if Immich's way to install an maintain stays at the current level it will not be accessible for the broad majority.

As for many Linux systems with the SNAP package I guess this goal is achieved, that worked just fine - at least for me.

Windows and Mac are a whole new level. Both could be done via a postgres DB, ngnix/ appache - plus ML parts and the "immich part" - I have no idea how complicated that would be - but that would be the "gold standard", GUI installer, set passwords for DB, set web-port - done!

The second to best option would probably be a VM as an image which one can pop into hyper-V, install it and setting passwords - I guess most ppl who would want to install a self hosted solution anyway would be able to do it if there is a documentation for it.

I am by no means a tech specialist but I do a lot of IT work for over 20 years (mostly hyper V) and do not get along with docker, I never needed it until I wanted to try Immich and I did not get it to work with windows - yeah dumb me maybe.... - however it was easy to fire up a VM in Hyper-V, takes just 10 clicks, and install Ubuntu 24.04 and install Immich via snap.

**Reply by 4v3ngR** [reply]
I spent the weekend trying to do a GUI installation (with platypus) and Unfortunately I was not able to get it working well. The biggest issue is that (for my script anyway) is that some tasks need to be done as the current user, some need to be done as root, and some need to be done as the immesh user. I was utilising the sudo command to handle the user changes, but that's not supported in scripts run through platypus.

There's the possibility of using apple scripts (oascript) to utilise the same functionality, so I might do further investigation into an automator app.

**Reply by 4v3ngR** [reply]
Just a quick update... I've released a macOS pkg installer that downloads and builds immich download it here: https://github.com/4v3ngR/immich-native-macos/releases

**Reply by blublub** [reply]
thats pretty cool! - sadly I have no mac to test it ;-(

**Reply by wuangyalin** [reply]
> Just a quick update... I've released a macOS pkg installer that downloads and builds immich download it here: https://github.com/4v3ngR/immich-native-macos/releases

Thanks for your sharing, just want to know if your pkg file working on the latest macos system with M chip mac?

**Comment by loeeeee** [comment]
Based on [Arter's scripts](https://github.com/arter97/immich-native), I made a [script](https://github.com/loeeeee/immich-in-lxc) that does the same thing but also supports CUDA, or hardware-accelerated video transcoding and machine-learning.

Leaving it here so that other people can find it useful.

**Comment by zebrapurring** [comment]
Hi all, for those interested in the macOS release I forked [4v3ngR's scripts](https://github.com/4v3ngR/immich-native-macos) and adapted them to generate a precompiled pkg file for Apple Silicon that can be installed directly.

The new repository is https://github.com/zebrapurring/immich-native-macos, and you can find there the release for [Immich v1.121.0](https://github.com/zebrapurring/immich-native-macos/releases/tag/v1.121.0).

**Comment by lingfish** [comment]
Firstly, thanks so much for this project, very impressive, and I (and I'm sure others) are really appreciative of your efforts.

Throwing my 2c in here too; I'm personally not much of a fan of docker, for many, many reasons; I was just about to attempt the remote machine learning capability (as my main desktop is a beast with a beast GPU in it), only to discover that install method requires docker too.

I *really* don't like how docker adds to and pollutes the network stack, and to install it just to get HW accel is not appealing.

**Reply by bo0tzz** [reply]
If the network handling is what bothers you, you might find podman more acceptable.

**Reply by dpantel** [reply]
What if it’s the 2+GB of docker crud I have to purge with every update?

**Reply by bo0tzz** [reply]
That sounds to me like just complaining for the sake of it. 2GB is not much, and the cleanup is trivially automated.

**Reply by dpantel** [reply]
What a silly reply.

Its GBs of downloads regularly and GBs of disk writes regularly, even if the updates are minor.

That may be trivial to you, but it’s not to me.

**Reply by lingfish** [reply]
(Well, this sub-thread went off the rails a bit!)

> If the network handling is what bothers you, you might find podman more acceptable.

Thanks for that suggestion; never looked into podman, but will.

That said, I too would still like to see non-docker options.

**Comment by gnull** [comment]
Native deployments are already supported in [NixOS](https://wiki.nixos.org/wiki/Immich). Immich with all dependencies (including ML server) install and run with just a few Nix config lines. NixOS provides pre-built packages for everything. It allows configuring any parameter of Immich, although the Immich service was added to NixOS quite recently and may lack documentation in some places.

I understand that what I wrote doesn't answer this feature request fully, but could hopefully help implement it.

**Reply by bo0tzz** [reply]
There's also an Arch AUR package. Note that any of these methods are not supported by us.

**Comment by anaxmedia** [comment]
Any chance this might be considered after the stable release this year? Would love a supported way to install in an LXC container.

**Reply by mmomjian** [reply]
No, we don’t plan on adding this in the future.

**Reply by MickLesk** [reply]
Sorry, I have already told people several times that we will not deploy Immich natively on LXC's. I know the processes here and the sometimes major changes, we are breaking our necks with this. We have made Dockge available for this purpose. Unfortunately, we also get the request almost weekly.

- Maintainer of community-scripts ORG (https://github.com/community-scripts/ProxmoxVE)

**Comment by syl779** [comment]
That's a pity. Was hoping to use my GPU but can't pass it through to a VM running docker, as I need it in my other LXCs (which can share it).

**Reply by anaxmedia** [reply]
I run docker in an LXC and passthrough the same GPU from my Plex LXC, works like a charm and has been stable for over a year.

**Comment by jonylentzmc** [comment]
+1 for dockerless install

**Comment by yagarea** [comment]
I am selfhosting immich and jellyfin and jellyfin is much easier to maintain and update because they have [official debian package](https://jellyfin.org/docs/general/installation/linux/#debuntu-debian-ubuntu-and-derivatives-using-apt). So you can just add their repo to sources and install it using apt and manage using systemd.

Creating debian package even without adding it to debian/ubuntu distributions could dramatically help increase adoption of immich.

**Reply by e64462** [reply]
I'm not sure that adoption is a high-priority goal.
